// fabric/dist/fabric.mjs
import { config } from "./src/config.mjs";
import { getEnv, getFabricDocument, getFabricWindow, setEnv } from "./src/env/index.mjs";
import { cache } from "./src/cache.mjs";
import { VERSION, iMatrix } from "./src/constants.mjs";
import { classRegistry } from "./src/ClassRegistry.mjs";
import { runningAnimations } from "./src/util/animation/AnimationRegistry.mjs";
import { Observable } from "./src/Observable.mjs";
import { Point } from "./src/Point.mjs";
import { createCollectionMixin } from "./src/Collection.mjs";
import { StaticCanvasDOMManager } from "./src/canvas/DOMManagers/StaticCanvasDOMManager.mjs";
import { StaticCanvas } from "./src/canvas/StaticCanvas.mjs";
import { Color } from "./src/color/Color.mjs";
import { Shadow } from "./src/Shadow.mjs";
import { Intersection } from "./src/Intersection.mjs";
import { FabricObject } from "./src/shapes/Object/Object.mjs";
import { Control } from "./src/controls/Control.mjs";
import { InteractiveFabricObject } from "./src/shapes/Object/InteractiveObject.mjs";
import { FabricObject as FabricObject$1 } from "./src/shapes/Object/FabricObject.mjs";
import { Rect } from "./src/shapes/Rect.mjs";
import { LayoutStrategy } from "./src/LayoutManager/LayoutStrategies/LayoutStrategy.mjs";
import { FitContentLayout } from "./src/LayoutManager/LayoutStrategies/FitContentLayout.mjs";
import { LayoutManager } from "./src/LayoutManager/LayoutManager.mjs";
import { Group } from "./src/shapes/Group.mjs";
import { util_exports } from "./src/util/index.mjs";
import { CanvasDOMManager } from "./src/canvas/DOMManagers/CanvasDOMManager.mjs";
import { controls_exports } from "./src/controls/index.mjs";
import { Canvas } from "./src/canvas/Canvas.mjs";
import { Gradient } from "./src/gradient/Gradient.mjs";
import { Pattern } from "./src/Pattern/Pattern.mjs";
import { BaseBrush } from "./src/brushes/BaseBrush.mjs";
import { Path } from "./src/shapes/Path.mjs";
import { PencilBrush } from "./src/brushes/PencilBrush.mjs";
import { Circle } from "./src/shapes/Circle.mjs";
import { CircleBrush } from "./src/brushes/CircleBrush.mjs";
import { SprayBrush } from "./src/brushes/SprayBrush.mjs";
import { PatternBrush } from "./src/brushes/PatternBrush.mjs";
import { Line } from "./src/shapes/Line.mjs";
import { Triangle } from "./src/shapes/Triangle.mjs";
import { Ellipse } from "./src/shapes/Ellipse.mjs";
import { Polyline } from "./src/shapes/Polyline.mjs";
import { Polygon } from "./src/shapes/Polygon.mjs";
import { FabricText } from "./src/shapes/Text/Text.mjs";
import { IText } from "./src/shapes/IText/IText.mjs";
import { Textbox } from "./src/shapes/Textbox.mjs";
import { ClipPathLayout } from "./src/LayoutManager/LayoutStrategies/ClipPathLayout.mjs";
import { FixedLayout } from "./src/LayoutManager/LayoutStrategies/FixedLayout.mjs";
import { ActiveSelection } from "./src/shapes/ActiveSelection.mjs";
import { Canvas2dFilterBackend } from "./src/filters/Canvas2dFilterBackend.mjs";
import { WebGLFilterBackend } from "./src/filters/WebGLFilterBackend.mjs";
import { getFilterBackend, initFilterBackend, setFilterBackend } from "./src/filters/FilterBackend.mjs";
import { FabricImage } from "./src/shapes/Image.mjs";
import { parseSVGDocument } from "./src/parser/parseSVGDocument.mjs";
import { loadSVGFromString } from "./src/parser/loadSVGFromString.mjs";
import { loadSVGFromURL } from "./src/parser/loadSVGFromURL.mjs";
import { isPutImageFaster, isWebGLPipelineState } from "./src/filters/utils.mjs";
import { filters_exports } from "./src/filters/filters.mjs";
export { ActiveSelection, BaseBrush, FabricObject as BaseFabricObject, Canvas, Canvas2dFilterBackend, CanvasDOMManager, Circle, CircleBrush, ClipPathLayout, Color, Control, Ellipse, FabricImage, FabricObject$1 as FabricObject, FabricText, FitContentLayout, FixedLayout, Gradient, Group, IText, FabricImage as Image, InteractiveFabricObject, Intersection, LayoutManager, LayoutStrategy, Line, FabricObject$1 as Object, Observable, Path, Pattern, PatternBrush, PencilBrush, Point, Polygon, Polyline, Rect, Shadow, SprayBrush, StaticCanvas, StaticCanvasDOMManager, FabricText as Text, Textbox, Triangle, WebGLFilterBackend, cache, classRegistry, config, controls_exports as controlsUtils, createCollectionMixin, filters_exports as filters, getEnv, getFabricDocument, getFabricWindow, getFilterBackend, iMatrix, initFilterBackend, isPutImageFaster, isWebGLPipelineState, loadSVGFromString, loadSVGFromURL, parseSVGDocument, runningAnimations, setEnv, setFilterBackend, util_exports as util, VERSION as version };
